### DISCUSSION TITLE
Use javascript syntax for regex

### DISCUSSION THREAD

**Discussion by svpace** [cue-lang/cue#1812] [Ideas]
Hi,

I would like to suggest to adopt javascript's syntax for regular expressions 

`=~ "[a-z0-9]*" => /[a-z0-9]*/`

Since json is a subset of javascript and cue is a superset of json I thing that it would make sense to reuse a familiar syntax instead of introducing a new one.

Best

**Comment by verdverm** [comment]
CUE uses Go's regexp library, which itself uses the widely adopted RE2 syntax

https://pkg.go.dev/regexp | https://github.com/cue-lang/cue/blob/master/pkg/regexp/regexp.go

> The syntax of the regular expressions accepted is the same general syntax used by Perl, Python, and other languages. More precisely, it is the syntax accepted by RE2 and described at https://golang.org/s/re2syntax

Another consideration would be the "does not match" operator `!~`, how would that look?

**Reply by svpace** [reply]
Sorry if I didn't made myself clear. My suggestion is not about the "inner" regex syntax (which I believe also is some PCRE dialect) but about the "outer" syntax which cue uses to represent regex literals.

A better example probably is `=~ "your regex" => /your regex/`. 

The syntax also allow for flags in the regex. For instance for case insensitive match `/regex/i`. Which flags are useful or viable in cue are open to discussion.

Regarding the `!=` operator, there's nothing in javascript about that. But I guess we could keep using the exclamation mark `!/regex/`. 

IMHO it kind of fits nicely with https://github.com/cue-lang/cue/issues/339#issue-936224918 `[X= =~"^P{Lu}"]: => [X=/^P{Lu}/]:`

WDYT?

**Reply by verdverm** [reply]
Having let it sit, I'm warming up to the idea. I'll leave the expert opines to @mpvl & @myitcv 

One thing to note is that you can use regular strings (`"..."`) and raw strings (`#"..."#`) and the resulting regexp has slightly different internal syntax. In other words, the raw string version has different escaping mechanisms. Not sure if they are using the same underlying implementations. I think there are 3 ways to get into regexp code (the two in the first sentence and the pkg in CUE's stdlib)

**Reply by mvdan** [reply]
Worth noting that Go's regular expression syntax also allows for flags; see https://pkg.go.dev/regexp/syntax.

Is this purely about the syntax being more familiar to JS developers, then?

**Reply by svpace** [reply]
Mostly yes.

I would argue that JS syntax for flags is easier and,  overall, the slash-notation makes regexes more explicit and unambiguous. But I am, obviously, biased on this front.

Nonetheless, given the subset/superset relationship between cue, json and js it makes sense to draw from js syntax.

**Reply by myitcv** [reply]
> Nonetheless, given the subset/superset relationship between cue, json and js it makes sense to draw from js syntax. 

I think there's just as strong an argument for using the Go reference point, given CUE builtins are based on Go's standard library. But I take your point.

However, I'm not clear we want/need to move away from representing regular expressions as strings, because otherwise we would need to support interpolation within `/.../`, which would cause us to reinvent the support we already have for strings.
